sklearn/metrics/_parallel_reductions.pyx

```diff
@@ -0,0 +1,1361 @@
+                    if Y_chunk_idx == self.Y_n_chunks - 1 and self.Y_n_samples_rem > 0:
+                        Y_end = Y_start + self.Y_n_samples_rem
+                    else:
+                        Y_end = Y_start + self.Y_n_samples_chunk
+
+                    self._reduce_on_chunks(
+                        X_start, X_end,
+                        Y_start, Y_end,
+                        thread_num,
+                    )
+
+                # Adjusting thread datastructures on the full pass on Y
+                self._on_X_prange_iter_finalize(thread_num, X_start, X_end)
+
+            # end: for X_chunk_idx
+
+            # Deallocating thread datastructures
+            self._on_X_parallel_finalize(thread_num)
+
+        # end: with nogil, parallel
+        return
+
+    @final
+    cdef void _parallel_on_Y(self) nogil:
+        """Computes the reduction of each vector (row) of X on Y
+        by parallelizing computation on chunks of Y.
+
+        Private datastructures are modified internally by threads.
+
+        Private template methods can be implemented on subclasses to
+        interact with those datastructures at various stages.
+        """
+        cdef:
+            ITYPE_t Y_start, Y_end, X_start, X_end, X_chunk_idx, Y_chunk_idx
+            ITYPE_t num_threads = min(self.Y_n_chunks, self.effective_omp_n_thread)
+            ITYPE_t thread_num
+
+        # TODO: put the "with nogil, parallel"-context here
+        # Allocating datastructures
+        self._on_Y_init(num_threads)
+
+        for X_chunk_idx in range(self.X_n_chunks):
+            X_start = X_chunk_idx * self.X_n_samples_chunk
+            if X_chunk_idx == self.X_n_chunks - 1 and self.X_n_samples_rem > 0:
+                X_end = X_start + self.X_n_samples_rem
+            else:
+                X_end = X_start + self.X_n_samples_chunk
+
+            with nogil, parallel(num_threads=num_threads):
+                thread_num = _openmp_thread_num()
+
+                # Initializing datastructures used in this thread
+                self._on_Y_parallel_init(thread_num)
+
+                for Y_chunk_idx in prange(self.Y_n_chunks, schedule='static'):
+                    Y_start = Y_chunk_idx * self.Y_n_samples_chunk
+                    if Y_chunk_idx == self.Y_n_chunks - 1 \
+                            and self.Y_n_samples_rem > 0:
+                        Y_end = Y_start + self.Y_n_samples_rem
+                    else:
+                        Y_end = Y_start + self.Y_n_samples_chunk
+
+                    self._reduce_on_chunks(
+                        X_start, X_end,
+                        Y_start, Y_end,
+                        thread_num,
+                    )
+                # end: prange
+
+            # Synchronizing the thread datastructures with the main ones
+            # This can potentially block
+            self._on_Y_after_parallel(num_threads, X_start, X_end)
+            # end: with nogil, parallel
```

Comment:
```suggestion
            # end: with nogil, parallel
            # Synchronizing the thread datastructures with the main ones
            # This can potentially block
            self._on_Y_after_parallel(num_threads, X_start, X_end)

```